lib/src/license_detection/lcs.dart

```diff
@@ -91,22 +123,63 @@ Iterable<TokenPair> _dynamicLcs({
     }
   }
 
-  // backtrack to construct the sequence
-  final matchesBackwards = <TokenPair>[];
+  // backtrack to construct the operations
+  final opsBackwards = <TokenOp>[];
   var i = m;
   var j = n;
 
-  while (i > 0 && j > 0) {
-    if (unknown[i - 1].value == known[j - 1].value) {
-      // building backwards, will need to reverse the list
-      matchesBackwards.add((unknown: unknown[i - 1], known: known[j - 1]));
-      i--;
-      j--;
-    } else if (table[i - 1][j] > table[i][j - 1]) {
-      i--;
+  while (i > 0 || j > 0) {
+    // shortcut: only deletes left
+    if (i == 0) {
+      opsBackwards.add(DeleteOp(known.take(j).toList()));
+      break;
+    }
+    // shortcut: only inserts left
+    if (j == 0) {
+      opsBackwards.add(InsertOp(unknown.take(i).toList()));
+      break;
+    }
+
+    bool isMatch() => i > 0 && j > 0 && table[i][j] == table[i - 1][j - 1] + 1;
+    bool isInsert() => i > 0 && table[i - 1][j] >= table[i][j - 1]; // j != 0
+    bool isDelete() => j > 0 && table[i - 1][j] < table[i][j - 1]; // i != 0
+
+    if (isMatch()) {
+      // match found
+      final matchPairsBackwards = <TokenPair>[];
+      do {
+        matchPairsBackwards.add((unknown: unknown[i - 1], known: known[j - 1]));
+        i--;
+        j--;
+      } while (isMatch());
+      opsBackwards.add(MatchOp(matchPairsBackwards.reverseIfNeeded()));
+    } else if (isInsert()) {
+      // insert into unknown
+      final insertTokens = <Token>[];
+      do {
+        insertTokens.add(unknown[i - 1]);
+        i--;
+      } while (isInsert() && !isMatch());
+      opsBackwards.add(InsertOp(insertTokens.reverseIfNeeded()));
     } else {
-      j--;
+      // delete from known
+      final deleteTokens = <Token>[];
+      do {
+        deleteTokens.add(known[j - 1]);
+        j--;
+      } while (isDelete() && !isMatch());
+      opsBackwards.add(DeleteOp(deleteTokens.reverseIfNeeded()));
+    }
+  }
+
+  return opsBackwards.reverseIfNeeded();
+}
+
+extension _ListExt<T> on List<T> {
+  List<T> reverseIfNeeded() {
```

Comment:
Consider just always reversing.
It doesn't seem meaningful to save the overhead here...

Reply:
I did also consider to do in-place replacing the items in the list, but wanted to start with a simpler code. I think it is not worth it for the coverage calculation, but may be worth it when we replace the text-diff with the token-diff.